File: src/main/java/com/jore/epoc/bo/orders/IncreaseProductivityOrder.java
package com.jore.epoc.bo.orders;

import org.hibernate.annotations.CompositeType;

import com.jore.datatypes.money.Money;
import com.jore.epoc.bo.accounting.FinancialAccounting;
import com.jore.epoc.bo.message.MessageLevel;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;

@Entity
public class IncreaseProductivityOrder extends AbstractSimulationOrder {
    @AttributeOverride(name = "amount", column = @Column(name = "increase_productivity_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "increase_productivity_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money increaseProductivityAmount;

    @Override
    public void execute() {
        if (getCompany().getAccounting().checkFunds(increaseProductivityAmount, getExecutionMonth().atEndOfMonth())) {
            setProductivityFactor();
            book(getExecutionMonth().atDay(1), "Increase productivity", FinancialAccounting.SERVICES, FinancialAccounting.BANK, increaseProductivityAmount);
            addMessage(MessageLevel.INFORMATION, "IncreaseProductivity", getExecutionMonth(), increaseProductivityAmount);
            setExecuted(true);
        } else {
            addMessage(MessageLevel.WARNING, "NoIncreaseInProductivityDueToFunding", getExecutionMonth(), increaseProductivityAmount, getCompany().getAccounting().getBankBalance(getExecutionMonth().atEndOfMonth()));
        }
    }

    @Override
    public int getSortOrder() {
        return 7;
    }

    public void setAmount(Money increaseProductivityAmount) {
        this.increaseProductivityAmount = increaseProductivityAmount;
    }

    private void setProductivityFactor() {
        Money pricePerPercentPoint = getCompany().getSimulation().getSettings().getPricePerPercentPointProductivity();
        double increase = increaseProductivityAmount.divide(pricePerPercentPoint).doubleValue();
        increase /= 100;
        getCompany().setProductivityFactor(getCompany().getProductivityFactor() + increase);
    }
}

File: EpocServer/Sources/App/Models/Orders/IncreaseProductivityOrder.swift
import Foundation

public final class IncreaseProductivityOrder: SimulationOrder {
    public var id: Int64?
    public var executionMonth: YearMonth = .of(2020, 1)
    public var isExecuted: Bool = false
    public weak var company: Company!
    public var increaseProductivityAmount: Money = Money.of("CHF", 0)

    public var sortOrder: Int { 7 }
    public var type: String { "Increase productivity" }

    public func getAmount() -> Money { increaseProductivityAmount }

    public func execute() {
        if company.accounting.checkFunds(increaseProductivityAmount, valueDate: executionMonth.atEndOfMonth()) {
            setProductivityFactor()
            book(bookingDate: executionMonth.atDay(1), bookingText: "Increase productivity",
                 debitAccount: FinancialAccounting.SERVICES, creditAccount: FinancialAccounting.BANK, amount: increaseProductivityAmount)
            addMessage(.information, "IncreaseProductivity", executionMonth, increaseProductivityAmount)
            isExecuted = true
        } else {
            addMessage(.warning, "NoIncreaseInProductivityDueToFunding", executionMonth, increaseProductivityAmount, company.accounting.getBankBalance(executionMonth.atEndOfMonth()))
        }
    }

    private func setProductivityFactor() {
        let pricePerPercentPoint = company.simulation.settings!.getPricePerPercentPointProductivity()
        var increase = increaseProductivityAmount.divide(pricePerPercentPoint).doubleValue
        increase /= 100
        company.productivityFactor += increase
    }
}
